### ISSUE TITLE
sequence or mapping pattern on generic variables makes case unreachable

### ISSUE THREAD

**Issue by pinterior** [microsoft/pyright#10689]
**Describe the bug**

Applying sequence or mapping pattern on a generic variable makes that case unreachable, and false positive `reportUnnecessaryComparison` is reported (if enabled).

**Code or Screenshots**

[pyright-play.net](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4RQKZgPZSQVQDtCMBjDAZwoEMoIBhbAWzFpgu0IF4koBXDAChBAEwwAzEOIDaAFQC6ACgAeaWQEoUgkDpBNqSUghCrtu86WoUMIAPoAaO1vMvdYWISSKARNYCO3uo6oCLsYAA21BAYIiBWIHyEmNRG1ABG4UKuOpbWIADe3gDW3mi2AL7O2ebuMJ4%2B%2BmCBwSChFBFRMXEUCUkYKQjpmWauuTa2VdU6rFTC4oqKAIz2AEzq6oLzhSVo3gBu3uXqQA)

```python
# pyright: reportUnnecessaryComparison=true

def f[T](x: T):
    match x:
        case _, _:
            print("seq")  # displayed as unreachable
        case {"k": _}:
            print("map")  # displayed as unreachable
        case _:
            pass

f((1,2))
f({"k": "v"})
```

```
$ pyright ma.py 
ma.py
  ma.py:3:13 - warning: TypeVar "T" appears only once in generic function signature
    Use "object" instead (reportInvalidTypeVarUse)
  ma.py:5:14 - error: Pattern will never be matched for subject type "T@f" (reportUnnecessaryComparison)
  ma.py:7:14 - error: Pattern will never be matched for subject type "T@f" (reportUnnecessaryComparison)
2 errors, 1 warning, 0 informations 
```

```
$ python ma.py 
seq
map
```

**VS Code extension or command-line**

command-line tool, 1.1.403
